### PR TITLE
Import/mbedtls 3.4.0 take3

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Take3 is to use `mbedtls_rsa_deduce_primes()` and `mbedtls_rsa_deduce_crt()` to deduce eventual missing fields.

### CODE REVIEW

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 1]
```diff
@@ -986,26 +992,33 @@ create_ec_priv_key_hidden_attributes(struct obj_attrs **out,
...
-			      &key_pair.grp.G, NULL, NULL);
+	ret = mbedtls_ecp_export(&key_pair, &key_pair_grp, &key_pair_d,
+				 &key_pair_Q);
+	if (ret) {
+		EMSG("Failed to export key");
+		goto out;
+	}
+
+	ret = mbedtls_ecp_mul(&key_pair_grp, &key_pair_Q, &key_pair_d,
+			      &key_pair_grp.G, NULL, NULL);
```
Comment (larper2axis) [inline ta/pkcs11/src/pkcs11_attributes.c]: The f_rng parameter to mbedtls_ecp_mul  is mandatory in 3.4.

File: ta/pkcs11/src/pkcs11_attributes.c [hunk 2]
```diff
@@ -2495,7 +2511,8 @@ static enum pkcs11_rc set_private_key_data_rsa(struct obj_attrs **head,
 	mbedtls_mpi_init(&p);
 	mbedtls_mpi_init(&q);
 
-	mbedtls_rc = mbedtls_pk_parse_key(&pk, data, key_size, NULL, 0);
+	mbedtls_rc = mbedtls_pk_parse_key(&pk, data, key_size,
+					  NULL, 0, NULL, 0);
```
Comment (larper2axis) [inline ta/pkcs11/src/pkcs11_attributes.c]: Also here the f_rng parameter is mandatory for blinding.

**Review by larper2axis — CHANGES_REQUESTED** [review]
See my comments about missing f_rng parameters.

File: core/lib/libtomcrypt/mpi_desc.c [hunk 3]
```diff
@@ -604,7 +604,7 @@ static int rng_read(void *ignored __unused, unsigned char *buf, size_t blen)
 
 static int isprime(void *a, int b __unused, int *c)
```
Comment (larper2axis) [inline core/lib/libtomcrypt/mpi_desc.c]: Remove the __unused annotation

**Comment by jenswikl** [comment]
> See my comments about missing f_rng parameters.

Thanks, should be taken care of now.

File: lib/libutee/tee_api_arith_mpi.c [hunk 4]
```diff
@@ -896,7 +896,8 @@ int32_t TEE_BigIntIsProbablePrime(const TEE_BigInt *op,
 
 	get_mpi(&mpi_op, op);
 
-	rc = mbedtls_mpi_is_prime(&mpi_op, rng_read, NULL);
+	rc = mbedtls_mpi_is_prime_ext(&mpi_op, MAX(confidenceLevel, 80U),
```
Comment (larper2axis) [inline lib/libutee/tee_api_arith_mpi.c:899]: Remove the __unused annotation on the parameter confidenceLevel.

**Review by larper2axis — APPROVED** [review]

**Comment by jenswikl** [comment]
Comments addressed. @larper2axis do you want to add an Acked-by or Reviewed-by for some commits? OK if I squash in the updates?

**Comment by jforissier** [comment]
- "libmbedtls: fix cipher_wrap.c for NIST AES Key Wrap mode"
  "libmedtls: core: update to mbedTLS 3.4.0 API"
  "core: ltc: configure internal SHA-1 and SHA-224"
  "core: ltc: configure internal MD5"
 
  `Acked-By: Jerome Forissier <jerome.forissier@linaro.org>`
 
- "ta: pkcs11: update to mbedTLS 3.4.0 API" + fixups:
  
  The mbedtls_pk_parse_key() has two new paramters with the new 3.x API
 
  s/paramters/parameters/
 
  `Acked-By: Jerome Forissier <jerome.forissier@linaro.org>`
  
- "core: LTC mpi_desc.c: update to mbedTLS 3.4.0 API" and "libutee: update to mbedTLS 3.4.0 API"

  The change to mbedtls_mpi_is_prime_ext() is not mentioned in the commit descriptions.
 
  `Acked-By: Jerome Forissier <jerome.forissier@linaro.org>`

- "libmedtls: core: update to mbedTLS 3.4.0 API"
 
  "The AES API has changed enough to prevent an efficient implementation of
  crypto_aes_enc_block() so switch the one provided via LTC is used
  instead."

  -> "so switch to the one provided via LTC instead"

  `Acked-By: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jforissier** [comment]
Should we be concerned with the CI errors? You've branched off of master haven't you, so I guess everything should build fine?

**Comment by jenswikl** [comment]
This one is needed also https://github.com/jenswi-linaro/optee_test/pull/1

**Comment by jenswikl** [comment]
Squashed and tags applied. Note that "libutee: update to mbedTLS 3.4.0 API" doesn't have any A-B or R-B tag yet.

**Comment by jforissier** [comment]
> Squashed and tags applied. Note that "libutee: update to mbedTLS 3.4.0 API" doesn't have any A-B or R-B tag yet.

For that commit:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswikl** [comment]
Tag applied. Are we happy with this? Should I push this branch to the new branch `import/mbedtls-3.4.0`? So I can create a new PR with a squash commit to be merged to master. Or should I wait a little longer?

**Comment by jforissier** [comment]
> Tag applied. Are we happy with this? Should I push this branch to the new branch `import/mbedtls-3.4.0`? So I can create a new PR with a squash commit to be merged to master. Or should I wait a little longer?

I'd like to run a few tests from my release check list first if you don't mind. Hopefully by the end of the day.

**Comment by jenswikl** [comment]
Sure, no problem.

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (vexpress-qemu_armv8a, mbedtls)`

For the record: I checked out a clean QEMUv8 environment and used the following command to build and run the tests:
`make -j check CFG_CRYPTOLIB_NAME=mbedtls CFG_CRYPTOLIB_DIR=lib/libmbedtls check XTEST_ARGS="-x pkcs11_1021 -x pkcs11_1022 -x pkcs11_1023"`

The failures in the PKCS#11 tests are a known issue with the previous MBedTLS already (https://github.com/OP-TEE/optee_os/issues/5418).

**Comment by jenswikl** [comment]
Pushed to https://github.com/OP-TEE/optee_os/tree/import/mbedtls-3.4.0

**Pull request merged**